### PR TITLE
Export DOS applications

### PR DESCRIPTION
**Pull request by robyoung**
Export DOS application information into three CSV files for successful,
failed and discretionary applications.

This script assumes that `insert-dos-framework-results.py` has already
been run and will fail with an assertion error if it has not (when it
comes across a declaration that is not complete while the supplier is
not marked as failed.

The file of successful applicants contains supplier information, contact
details and counts for completed, failed and draft services.
The file of failed applicants contains supplier information, a list of
failed mandatory questions, contact details and counts for completed,
failed and draft services.
The file of discretionary applicants contains supplier information,
discretionary questions and their answers, contact details and counts
for completed, failed and draft services.

Service level failed questions are not included in these reports.

### CODE REVIEW

File: dmscripts/export_dos_suppliers.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+def export_suppliers(client, content_loader, output_dir):
+    records = find_suppliers_with_details(client, content_loader, FRAMEWORK_SLUG)
+
+    handlers = [SuccessfulHandler(), FailedHandler(), DiscretionaryHandler()]
+
+    with MultiCSVWriter(output_dir, handlers) as writer:
+        for i, record in enumerate(records):
+            writer.write_row(record)
+            if i % 100 == 0:
```
Comment (TheDoubleK) [inline dmscripts/export_dos_suppliers.py]: I like how we get an output line for the zeroth row :) - this confused me running against preview, but will be nice for when we run prod where there' s a lot more data.

File: dmscripts/export_dos_suppliers.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+
+def find_suppliers_with_details(client, content_loader, framework_slug):
+    pool = ThreadPool(30)
+
+    content_loader.load_manifest(framework_slug, 'declaration', 'declaration')
+    declaration_content = content_loader.get_manifest(framework_slug, 'declaration')
+
+    records = find_suppliers(client, framework_slug)
+    records = filter(None, records)
```
Comment (TheDoubleK) [inline dmscripts/export_dos_suppliers.py]: Not sure why we need this filter? I don't think the API will return any empty supplier IDs?

File: tests/test_export_dos_suppliers.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+import pytest
+import six
+from mock import Mock, call, patch, ANY, mock_open
+
+from dmscripts import export_dos_suppliers
+from dmscripts.insert_dos_framework_results import CORRECT_DECLARATION_RESPONSES
+from dmapiclient import HTTPError
+
+
+# TODO: move to conftest
```
Comment (TheDoubleK) [inline tests/test_export_dos_suppliers.py]: Is this TODO going to get done now, or some indeterminate time in the never future?

File: dmscripts/export_dos_suppliers.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+            if supplier_framework['declaration'].get('status') != 'complete':
+                assert supplier_framework['onFramework'] is False, \
+                    "Incomplete declaration but not failed, have you inserted the framework result?"
+
+            return dict(record,
+                        declaration=supplier_framework['declaration'],
+                        onFramework=supplier_framework['onFramework'])
+        except HTTPError as e:
+            if e.status_code != 404:
```
Comment (TheDoubleK) [inline dmscripts/export_dos_suppliers.py]: I'm not sure we'd want to fail silently for a 404 here as it shouldn't ever happen as far as I can think? The suppliers returned by `find_suppliers` should by definition be only those suppliers that have a `supplier_frameworks` entry for DOS. So a 404 here should probably set some alarm bells ringing.

File: tests/test_export_dos_suppliers.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        ('draft_user-research-participants', 0),
+    ]
+
+SUCCESSFUL_RECORD = {
+    'onFramework': True,
+    'supplier': {'id': 1,
+                 'name': 'Bluedice Ideas Limited'},
+    'declaration': {'nameOfOrganisation': 'Young Money Franchise LTD',
+                    'primaryContact': 'Yolo Swaggins',
+                    'primaryContactEmail': 'yolo.swagins@example.com'},
```
Comment (TheDoubleK) [inline tests/test_export_dos_suppliers.py]: Should `swagins` have a double `g` in the email address do you think?

**Comment by TheDoubleK** [comment]
Reviewed - a few minor comments but essentially all looks good to go. Sterling work on the tests.
I've sent outputs from preview to Andrew, and as long as he confirms that the counts, etc for the test accounts match up OK I reckon we can go ahead and run the scripts on Staging/Prod.

File: dmscripts/export_dos_suppliers.py (hunk 2 above)
Comment (robyoung) [inline dmscripts/export_dos_suppliers.py]: Good call, I was originally just listing all suppliers.

File: tests/test_export_dos_suppliers.py (hunk 3 above)
Comment (robyoung) [inline tests/test_export_dos_suppliers.py]: Good spot, it was a reminder to move it somewhere your tests and mine can both make use of it.

**Comment by TheDoubleK** [comment]
Updates look good, and Preview data all confirmed as correct.
:+1:

**Pull request merged**